### DISCUSSION TITLE
mgmt interface with c8000v

### DISCUSSION THREAD

**Discussion by nagraj-kulkarni** [ipspace/netlab#1303] [Q&A] [answered]
Hi, having hard time getting the connectivity from cisco 8000v to external world. Understand that the mgmt interface  i.e first interface , gets qemu dhcp 10.x.x.x and it gets discovered by external routers etc. But i am unable to get the external reachability from the 8000v router itself, tried using the mgmt,addressing schemes in topology/static routes but not helping. How do get it to ping say 8.8.8.8 from the router ? appreciate any help :)

**Comment by ipspace** [comment]
The management interface (the first Ethernet interface) of Cisco 8000v (and most other devices) is in the management VRF. It should get the IP address from the libvirt DHCP server and it should be in the 192.168.121.0/24 range. To reach the outside world through the management VRF, you have to use the VRF version of the commands (**ping vrf** or similar)

You can also connect one of the data-plane interfaces to the outside world. See https://netlab.tools/labs/libvirt/#connecting-to-the-outside-world for details (assuming you're using libvirt because you mentioned Qemu)

**Reply by nagraj-kulkarni** [reply]
Hello, i am using the clab provider and the first interface is getting ip in the subnet 10.0.0.x from qemu dhcp. There is no vrf configured. 
 I used "addressing/mgmt" etc in the topology file so that it gets the first interface ip from these subnets ( 192.x). netlab status shows the managment ip as 192.x network, but the show running-config in the router does not show any interface having ip in 192.x network,so i believe clab is doing a magic in the underlying bridge, to get reachaibility to router. So in bried,  I can do ssh to the router using 192.x, but no interface has ip  configured in 192.x ( as per show running), ping from router to external machine (where i am running netlab) does not work, i can not give the default route as 192.x.x.1 in the router as this network is unknown :)

**Comment by ssasso** [comment, marked as answer]
When using containerlab, with the vrnetlab containers, a sort of 1:1 NAT is performed in userspace by qemu.

You will find an interface with 10.0.0.something, and the default gateway for this is 10.0.0.2

**Comment by ipspace** [comment]
Ah, vrnetlab. I never understood why someone would use that when there's a perfectly fine way to build virtual machine Vagrant boxes without containers. The convoluted networking they had to do to get things working led me to write this in https://netlab.tools/labs/clab/#using-vrnetlab-containers

> vrnetlab is an independent open-source project. If it fails to produce a working container image ([example](https://github.com/hellt/vrnetlab/issues/231)), please contact them.

Anyway, vrnetlab got much better after @hellt forked it. At the very minimum, you should use the latest vrnetlab code from https://github.com/hellt/vrnetlab/ to build the container. For example, https://github.com/hellt/vrnetlab/tree/master/c8000v/docker puts the management interface into a VRF and configures the default route for the management VRF.

**Reply by nagraj-kulkarni** [reply]
Hello @ipspace  @ssasso  
used vrnetlab with hellt fork for csr1000v  docker image and observe a weird thing. when lab is built with multiple csr containers, all of the routers get the same ip and mac address on gi/1. Wondering how you people got rid of this issue while playing around !!  

leaf-1#show run int gi1
Building configuration...

Current configuration : 179 bytes
!
interface GigabitEthernet1
 vrf forwarding clab-mgmt
 ip address 10.0.0.15 255.255.255.0
 negotiation auto
 no lldp transmit
 no lldp receive
 no mop enabled
 no mop sysid
end

leaf-1#exit


on 2nd router ++++++++

leaf-2#show run int gi1
Building configuration...

Current configuration : 179 bytes
!
interface GigabitEthernet1
 vrf forwarding clab-mgmt
 ip address 10.0.0.15 255.255.255.0
 negotiation auto
 no lldp transmit
 no lldp receive
 no mop enabled
 no mop sysid
end

snmp shows same mac address on mgmt port ( gi 1) on two routers 

Router1--->  snmpwalk -v2c -c public 192.168.10.101 1.3.6.1.2.1.2.2.1.6.1 
iso.3.6.1.2.1.2.2.1.6.1 = Hex-STRING: C0 00 01 00 CA FE
Router2 ---> snmpwalk -v2c -c public 192.168.10.102 1.3.6.1.2.1.2.2.1.6.1 
iso.3.6.1.2.1.2.2.1.6.1 = Hex-STRING: C0 00 01 00 CA FE


used the below snippet for above topo:
addressing:
  loopback:
    ipv4: 10.255.0.0/24
  router_id:
    ipv4: 10.255.0.0/24



Without the above snippet, it looks like :

leaf-1#show run int gi1
Building configuration...

Current configuration : 207 bytes
!
interface GigabitEthernet1
 mac-address c000.0101.cafe
 vrf forwarding clab-mgmt
 ip address 10.0.0.15 255.255.255.0
 negotiation auto
 no lldp transmit
 no lldp receive
 no mop enabled
 no mop sysid
end

leaf-1#exit



On 2nd router ++++++++
leaf-2#show run int gi1
Building configuration...

Current configuration : 207 bytes
!
interface GigabitEthernet1
 mac-address c000.0102.cafe
 vrf forwarding clab-mgmt
 ip address 10.0.0.15 255.255.255.0
 negotiation auto
 no lldp transmit
 no lldp receive
 no mop enabled
 no mop sysid
end

**Reply by ssasso** [reply]
Hi @nagraj-kulkarni ,
That's not a problem, since the VM management interface is **local** to the vrnetlab container, and not exposed.

The qemu process takes care of NAT-ing that IP address to the real management one assigned by containerlab.

**Reply by nagraj-kulkarni** [reply]
@ssasso Understood :),  but snmp gives the same interface ip and macs and apps/nms might assume its single device and ignore other routers..

**Reply by ssasso** [reply]
@nagraj-kulkarni you need to modify vrnetlab in that case... 

You can override `gen_mgmt(self)` in https://github.com/hellt/vrnetlab/blob/master/csr/docker/launch.py

Ref:

https://github.com/hellt/vrnetlab/blob/afd1f54c9b735b784a281f8ce9778c0c6b868028/common/vrnetlab.py#L296

**Reply by ipspace** [reply]
Also, as I pointed out on September 3rd, you could run Cat8000v and CSR1000v as virtual machines and never experienced these challenges ;) #JustSaying

**Reply by nagraj-kulkarni** [reply]
hmm, reading about it..looks like i need to download serial qcow2 images, install vagrant and boxes, setup initial configs

**Reply by ipspace** [reply]
> looks like i need to download serial qcow2 images

If you built the container yourself, you already have the correct image.

> install vagrant

If you're on a recent Ubuntu, it's as easy as "netlab install ubuntu"

>  and boxes

"netlab libvirt package csr disk-image"

> setup initial configs

That's the only difference between vrnetlab and what netlab is doing for vagrant boxes. I didn't want to waste my time writing expect scripts to parse vendor chaff.

I'm not pushing you to do it, it just removes a layer of abstraction (vrnetlab container is running KVM inside anyway).

**Reply by hellt** [reply]
Containerlab/vrnetlab plans to introduce a mode where "real ip" is configured in the VMs mgmt vrf. 
Potentially this year

**Reply by ipspace** [reply]
@hellt: That's really nice to hear. It would be static (configured by containerlab), not DHCP-assigned, right?

Also, please make sure the MAC addresses are unique (or derived from the IP address ;)

**Reply by hellt** [reply]
It would be assigned by container runtime in this case. Meaning you don't have to statically type it in, but you could assign it statically as well

From the node config perspective this would be a static configuration of the mgmt IP

**Reply by nagraj-kulkarni** [reply]
@ipspace @ssasso is there a way to create vagrant box without bringing down other labs. I have few labs setup and dont want to disturb them. Getting below error  with netlab libvirt package csr disk-image:

netlab cannot create a Vagrant box while there are running labs. Inspect the
status of your lab instances with "netlab status" and stop them with
"netlab down" or "netlab status cleanup --all"

**Reply by ipspace** [reply]
Unfortunately not. Keep working with the setup you have right now, and build that box once you have to shutdown or reload the server.

**Reply by ssasso** [reply]
> Unfortunately not. Keep working with the setup you have right now, and build that box once you have to shutdown or reload the server.

Or create a virtual machine with nested virtualization capabilities just to build the Vagrant box ;)

(even on some public cloud, most probably it should take less than an hour to build the vagrant box, so in theory the cost should be affordable)

**Reply by nagraj-kulkarni** [reply]
ok Thank you

**Reply by routingpackets** [reply]
@hellt @ipspace Hi guys, is there any updates if the real IP is already possible to be configured on the mgmt interfaces of the devices? I am facing a similar behavior with the n9kv images. They all get 10.0.0.15 and is NATed.
Thanks in advance!

**Reply by ipspace** [reply]
AFAIK, the networking within the vrnetlab containers hasn't changed, but it got stable enough that we stopped tracking it; from the outside, everything just works. I would suggest you open an issue @ https://github.com/srl-labs/vrnetlab

**Comment by ipspace** [comment]
Anyway, it looks like I didn't stress enough the importance of using the (latest version of) @hellt fork of vrnetlab to build the container images, so I (hopefully) fixed the documentation: https://github.com/ipspace/netlab/commit/20991a5886ae682297293abd4880a1f6346f0fe4

**Reply by nagraj-kulkarni** [reply]
Thanks for the quick responses :)  for now, i pointed the default gw as suggested by Steffano and it helped. i was in such an awful situation :) will try with vrnetlab fork..Thanks again..

**Comment by ipspace** [comment]
Here's another idea to build that Vagrant box:

* Start a **ubuntu:22.04** container (making sure to map the current directory or some such into it)
* Connect to the container, run the netlab / ubuntu / libvirt installation procedure
* Build the Vagrant box within the container
* Ensure the box-building artifacts are in the host directory mapped into the container (and thus accessible from your server)
* Install Vagrant box on your server with **vagrant box add _json-file_** command

**Reply by nagraj-kulkarni** [reply]
Thanks @ipspace , able to create a box in this fashion, But when i bring the lab up, getting lot of errors !!

`┌──────────────────────────────────────────────────────────────────────────────────┐
│ CHECKING virtualization provider installation                                    │
└──────────────────────────────────────────────────────────────────────────────────┘
[SUCCESS] libvirt installed and working correctly

┌──────────────────────────────────────────────────────────────────────────────────┐
│ STARTING libvirt nodes                                                           │
└──────────────────────────────────────────────────────────────────────────────────┘
[CREATED] creating libvirt management network nl_mgmt_12
provider libvirt: executing vagrant up --provider libvirt
Bringing machine 'leaf1' up with 'libvirt' provider...
Bringing machine 'leaf2' up with 'libvirt' provider...
Bringing machine 'spine1' up with 'libvirt' provider...
Bringing machine 'host1' up with 'libvirt' provider...
Bringing machine 'host2' up with 'libvirt' provider...
==> leaf1: An error occurred. The error will be shown after all tasks complete.
==> spine1: An error occurred. The error will be shown after all tasks complete.
==> host1: An error occurred. The error will be shown after all tasks complete.
==> leaf2: An error occurred. The error will be shown after all tasks complete.
==> host2: An error occurred. The error will be shown after all tasks complete.
An error occurred while executing multiple actions in parallel.
Any errors that occurred are shown below.

An error occurred while executing the action on the 'leaf1'
machine. Please handle this error then try again:

There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The following settings shouldn't exist: allow_fstab_modification, allow_hosts_modification

Libvirt Provider:
* The following settings shouldn't exist: qemu_use_session



An error occurred while executing the action on the 'leaf2'
machine. Please handle this error then try again:

There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The following settings shouldn't exist: allow_fstab_modification, allow_hosts_modification

Libvirt Provider:
* The following settings shouldn't exist: qemu_use_session



An error occurred while executing the action on the 'spine1'
machine. Please handle this error then try again:

There are errors in the configuration of this machine. Please fix
the following errors and try again:

vm:
* The following settings shouldn't exist: allow_fstab_modification, allow_hosts_modification

Libvirt Provider:
* The following settings shouldn't exist: qemu_use_session
`

**Reply by ipspace** [reply]
This is unusual. What version of Vagrant and vagrant-libvirt provider are you using? You can display them with `vagrant --version` and `vagrant plugin list`

**Reply by nagraj-kulkarni** [reply]
root@219e261f47d8:/workspace# vagrant --version
Vagrant 2.4.0
root@219e261f47d8:/workspace# vagrant plugin list
vagrant-libvirt (0.12.2, global)
  - Version Constraint: > 0

**Reply by ipspace** [reply]
I have almost the same version and never encountered similar errors. The parameters are also perfectly valid according to Vagrant documentation https://developer.hashicorp.com/vagrant/docs/vagrantfile/machine_settings

I have no idea what's going on.

**Reply by nagraj-kulkarni** [reply]
guess i am the privileged person :)

**Reply by nagraj-kulkarni** [reply]
@ipspace @ssasso   on spine-leaf lab ( fabric: plugin) with csr docker image , see lldp neighbours keep changing. do you know of any known issues/caveats apart from MTU size and sub-interfaces support?    

leaf-1#show lldp ne
======================
Capability codes:
    (R) Router, (B) Bridge, (T) Telephone, (C) DOCSIS Cable Device
    (W) WLAN Access Point, (P) Repeater, (S) Station, (O) Other

Device ID             Local Intf     Hold-time  Capability      Port ID
spine-1.example  Gi2            120                    R               Gi2       

Total entries displayed: 1

leaf-1#show lldp ne
======================
Capability codes:
    (R) Router, (B) Bridge, (T) Telephone, (C) DOCSIS Cable Device
    (W) WLAN Access Point, (P) Repeater, (S) Station, (O) Other

Device ID             Local Intf     Hold-time  Capability      Port ID
spine-2.example  Gi3            120                     R               Gi2       

Total entries displayed: 1

leaf-1#


leaf-1#show lldp ne
==========================
Capability codes:
    (R) Router, (B) Bridge, (T) Telephone, (C) DOCSIS Cable Device
    (W) WLAN Access Point, (P) Repeater, (S) Station, (O) Other

Device ID           Local Intf     Hold-time  Capability      Port ID
spine-1.example  Gi2            120                  R               Gi2

Total entries displayed: 1



Topology.yaml =======> 

provider: clab
defaults.devices.csr.clab.image: "vrnetlab/vr-csr:17.03.08"
plugin: [ fabric ]
defaults.device: csr

fabric.spines: 2
fabric.leafs: 4

fabric.leaf.name: leaf-{count}
fabric.spine.name: spine-{count}

groups:
  leafs:
    module: [ ospf, bgp ]
    bgp.as: 65000
  spines:
    module: [ ospf ]

nodes:
  host-1:
    device: linux
    image: ubuntu:latest
  host-2:
    device: linux
    image: ubuntu:latest

links:
- host-1:
   leaf-1:
- host-2:
  leaf-2:

**Reply by ipspace** [reply]
A) It might be a good idea to open a new discussion for a completely unrelated issue
B) No idea. You could try CDP/LLDP debugging. Without knowing anything else, it looks like either the LLDP packets are getting lost or there's a timeout mismatch.

You could try the new packet capture capability added in 1.9.1 to figure out what's going on.

**Reply by ipspace** [reply]
Ran a few tests:

A) Contrary to what I said, you cannot do packet capture on vrnetlab containers because the outside container interface names don't match the inside names (https://github.com/ipspace/netlab/issues/1380)
B) I'm experiencing the exact same behavior. As I have plenty of CPU and can see the LLDP packets on both uplinks, it's not resource related. It looks like all CSRs report the same Chassis ID MAC address with different management IP addresses (loopback interfaces) and that might confuse the LLDP process.

I also checked the serial numbers, and it looks like all vrnetlab-based CSRs have the same serial number. Have no idea how to change that or the Chassis ID, it might be based on the MAC address of the first interface (and you know what the problem is there).

However, LLDP does work correctly with the CSR VMs:

```
leaf_1#show lldp neighbors
Capability codes:
    (R) Router, (B) Bridge, (T) Telephone, (C) DOCSIS Cable Device
    (W) WLAN Access Point, (P) Repeater, (S) Station, (O) Other

Device ID           Local Intf     Hold-time  Capability      Port ID
spine_1.lab.local   Gi2            120        R               Gi2
spine_2.lab.local   Gi3            120        R               Gi2

Total entries displayed: 2
```

You might try to change the vrnetlab code to make the MAC address random (or open an issue and ask someone else to do it).

**Reply by nagraj-kulkarni** [reply]
great, Thanks a lot @ipspace for checking on this and all the details. When i checked on the CSRs that they are waiting for lldp frame, was wondering how to capture the packet and check.
